cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,122 +1265,752 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
+                                                   accum_size);
+
+  } else {
+    // we instantiate an "iteration state" for the regular single pass contiguous_split
+    // consisting of 1 iteration with all of the batches and totalling `total_size` bytes.
+    auto const total_size = std::reduce(h_buf_sizes, h_buf_sizes + num_partitions);
+
+    // 1 iteration with the whole size
+    return std::make_unique<chunk_iteration_state>(
+      std::move(d_batched_dst_buf_info),
+      std::move(d_batch_offsets),
+      std::move(std::vector<std::size_t>{static_cast<std::size_t>(num_batches)}),
+      std::move(std::vector<std::size_t>{total_size}),
+      total_size);
   }
-  return result;
+}
+
+void copy_data(int num_batches_to_copy,
+               int starting_batch,
+               uint8_t const** d_src_bufs,
+               uint8_t** d_dst_bufs,
+               rmm::device_uvector<dst_buf_info>& d_dst_buf_info,
+               uint8_t* user_buffer,
+               rmm::cuda_stream_view stream)
+{
+  constexpr size_type block_size = 256;
+  if (user_buffer != nullptr) {
+    auto index_to_buffer = [user_buffer] __device__(unsigned int) { return user_buffer; };
+    copy_partitions<block_size><<<num_batches_to_copy, block_size, 0, stream.value()>>>(
+      index_to_buffer, d_src_bufs, d_dst_buf_info.data() + starting_batch);
+  } else {
+    auto index_to_buffer = [d_dst_bufs,
+                            dst_buf_info = d_dst_buf_info.data(),
+                            user_buffer] __device__(unsigned int buf_index) {
+      auto const dst_buf_index = dst_buf_info[buf_index].dst_buf_index;
+      return d_dst_bufs[dst_buf_index];
+    };
+    copy_partitions<block_size><<<num_batches_to_copy, block_size, 0, stream.value()>>>(
+      index_to_buffer, d_src_bufs, d_dst_buf_info.data() + starting_batch);
+  }
+}
+
+/**
+ * @brief Function that checks an input table_view and splits for specific edge cases.
+ *
+ * It will return true if the input is "empty" (no rows or columns), which means
+ * special handling has to happen in the calling code.
+ *
+ * @param input table_view of source table to be split
+ * @param splits the splits specified by the user, or an empty vector if no splits
+ * @returns true if the input is empty, false otherwise
+ */
+bool check_inputs(cudf::table_view const& input, std::vector<size_type> const& splits)
+{
+  if (input.num_columns() == 0) { return true; }
+  if (splits.size() > 0) {
+    CUDF_EXPECTS(splits.back() <= input.column(0).size(),
+                 "splits can't exceed size of input columns");
+  }
+  {
+    size_type begin = 0;
+    for (std::size_t i = 0; i < splits.size(); i++) {
+      size_type end = splits[i];
+      CUDF_EXPECTS(begin >= 0, "Starting index cannot be negative.");
+      CUDF_EXPECTS(end >= begin, "End index cannot be smaller than the starting index.");
+      CUDF_EXPECTS(end <= input.column(0).size(), "Slice range out of bounds.");
+      begin = end;
+    }
+  }
```

Comment:
Extra braces? Also, the for-loop can be simplified when the `i` iterator is not needed.
```suggestion
  size_type begin = 0;
  for (auto end : splits) {
    CUDF_EXPECTS(begin >= 0, "Starting index cannot be negative.");
    CUDF_EXPECTS(end >= begin, "End index cannot be smaller than the starting index.");
    CUDF_EXPECTS(end <= input.column(0).size(), "Slice range out of bounds.");
    begin = end;
  }
```